tensorflow/core/kernels/mkl_qmatmul_op.cc

```diff
@@ -0,0 +1,875 @@
+          __FILE__, ":", __LINE__);
+      OP_REQUIRES_OK(
+          context,
+          errors::Aborted("Operation received an exception:", error_msg));
+    }
+
+    // Compute additional outputs: min/max scalars.
+    const float min_input = context->input(3).flat<float>()(0);
+    const float max_input = context->input(4).flat<float>()(0);
+    const float min_weight = context->input(5).flat<float>()(0);
+    const float max_weight = context->input(6).flat<float>()(0);
+
+    float min_output_value;
+    float max_output_value;
+    if (std::is_same<Toutput, quint8>::value ||
+        std::is_same<Toutput, qint8>::value) {
+      // This is the case the inner-product and requantization are fused.
+      // "min_freezed_output" and "max_freezed_output" are the actual range
+      // for the output
+      min_output_value = context->input(7).flat<float>()(0);
+      max_output_value = context->input(8).flat<float>()(0);
+    } else {
+      MklQuantizationRangeForMultiplication<quint8, qint8, qint32>(
+          min_input, max_input, min_weight, max_weight, &min_output_value,
+          &max_output_value);
+    }
+
+    Tensor* output_min = nullptr;
+    Tensor* output_max = nullptr;
+    MklDnnShape output_min_mkl_shape, output_max_mkl_shape;
+    output_min_mkl_shape.SetMklTensor(false);
+    output_max_mkl_shape.SetMklTensor(false);
+    AllocateOutputSetMklShape(context, 1, &output_min, {},
+                              output_min_mkl_shape);
+    AllocateOutputSetMklShape(context, 2, &output_max, {},
+                              output_max_mkl_shape);
+    output_min->flat<float>()(0) = min_output_value;
+    output_max->flat<float>()(0) = max_output_value;
+  }
+
+ protected:
+  virtual void ExtendMklDnnMatMulFwdParams(OpKernelContext* context,
+                                           MklDnnMatMulFwdParams& params) {
+    // Append data type names of input, weight, bias, and output.
+    params.dtypes.append(typeid(Tinput).name());
+    params.dtypes.append(typeid(Tweight).name());
+    params.dtypes.append(typeid(Tbias).name());
+    params.dtypes.append(typeid(Toutput).name());
+
+    // When the output type is quint8, the output data is requantized into
+    // quint8. A post_op "output_scale" is added to do the conversion.
+    if (std::is_same<Toutput, quint8>::value ||
+        std::is_same<Toutput, qint8>::value) {
+      const float min_input = context->input(3).flat<float>()(0);
+      const float max_input = context->input(4).flat<float>()(0);
+      const float min_weight = context->input(5).flat<float>()(0);
+      const float max_weight = context->input(6).flat<float>()(0);
+      const float min_freezed_output = context->input(7).flat<float>()(0);
+      const float max_freezed_output = context->input(8).flat<float>()(0);
+
+      float min_output_value;
+      float max_output_value;
+      MklQuantizationRangeForMultiplication<quint8, qint8, qint32>(
```

Comment:
The code to calculate `min_output_value` and `max_output_value` is the same as the one in `Compute` (line 576). Could you please refactor it?

Reply:
@penpornk Refactored in the latest commit.